### PR TITLE
Add touch button client and OTA logging

### PR DESCRIPTION
**Pull request by rytrose**

### CODE REVIEW

File: arduino/WandCore/CAPClient.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+*/
+
+#include "CAPClient.h"
+
+// touch threshold
+const int touchThold = 50; //above means not touched
+
+// Touch surface timing variables
+int debounce = 20;                // ms debounce period to prevent flickering when pressing or releasing the button
+int DCgap = 250;                  // max ms between clicks for a double click event
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: IDK about you, but I don't really like the `DC*` nomenclature, a the very least make it `DT` since "click" doesn't make sense any more, but I'd prefer `doubleClickGap` (I generally prefer detail in variable names). The comment could be updated, too.

File: arduino/WandCore/CAPClient.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+int holdTime = 1000;              // ms hold period: how long to wait for tap+hold event
+
+// Touch surface variables
+boolean isTouched = false;        // value read from touch surface
+boolean wasTouchedLast = false;   // buffered value of the button's previous state
+boolean DCwaiting = false;        // whether we're waiting for a double tap (down)
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: What does `(down)` mean here?

File: arduino/WandCore/CAPClient.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+boolean DConUp = false;           // whether to register a double tap on next release, or whether to wait and tap
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: This comment is also confusing without looking at how it's being used.

File: arduino/WandCore/CAPClient.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#include "Arduino.h"
+
+/* checkTouchSurface
+ * Takes a pin attached to a touch surface
+ * Returns an int representing the UI action a user has taken with the touch surface 
+ * 
+ * 0 = not interacting with the surface
+ * 1 = user single tapped
+ * 2 = user double tapped
+ * 3 = user tapped and is holding
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.h]: I would `define` these as constants, it's hard to read the code without being able to refer to them by name.
e.g.
```
#define NOT_TOUCHING 0
#define SINGLE_TAP 1
...
```

File: arduino/WandCore/CAPClient.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+boolean singleOK = true;          // whether it's OK to do a single tap
+unsigned long downTime = -1;      // time the surface was tapped
+unsigned long upTime = -1;        // time the surface was released
+boolean ignoreUp = false;         // whether to ignore the surface release because the tap+hold was triggered
+boolean waitForUp = false;        // when held, whether to wait for the up event
+boolean holdEventPast = false;    // whether or not the hold event happened already
+
+bool checkIfTouched(int pin) {
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: I wouldn't have this or the other function take the pin as input, rather define a `setupTouchButtonClient(int pin)` function to set the pin in `setup()`, like our other clients do.

File: arduino/WandCore/CAPClient.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
+/*
+  Capacitive Touch Client
+
+  Manages a Touch Surface Input
+*/
+
+#include "CAPClient.h"
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: I'm not a huge fan of `CAPClient`, what about `TouchButtonClient`?

File: arduino/WandCore/CAPClient.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  
+  // Surface is just touched
+  if (isTouched == true && wasTouchedLast == false && (millis() - upTime) > debounce)
+  {
+    downTime = millis();
+    ignoreUp = false;
+    // waitForUp = false; // may be unnecessary?
+    singleOK = true;
+    holdEventPast = false;
+    if ((millis() - upTime) < DCgap && DConUp == false && DCwaiting == true)  DConUp = true;
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: bruh help a brother out with a comment

File: arduino/WandCore/CAPClient.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  else return true;
+}
+
+int getTouchEvent(int pin) {   
+  int event = 0;
+  isTouched = checkIfTouched(pin);
+  
+  // Surface is just touched
+  if (isTouched == true && wasTouchedLast == false && (millis() - upTime) > debounce)
+  {
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: I personally don't like conditional blocks having the brace start on a new line, and we don't do it in our other code.

File: arduino/WandCore/CAPClient.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        event = 2;
+        DConUp = false;
+        DCwaiting = false;
+        singleOK = false;
+      }
+    }
+  }
+  
+  // Test for single tap event: DCgap expired
+  if (isTouched == false && (millis() - upTime) >= DCgap && DCwaiting == true && DConUp == false && singleOK == true && event != 2)
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: For all these crazy conditionals, I would love it if each expression was on its own line with an english comment describing each.

File: arduino/WandCore/CAPClient.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+int getTouchEvent(int pin) {   
+  int event = 0;
+  isTouched = checkIfTouched(pin);
+  
+  // Surface is just touched
+  if (isTouched == true && wasTouchedLast == false && (millis() - upTime) > debounce)
+  {
+    downTime = millis();
+    ignoreUp = false;
+    // waitForUp = false; // may be unnecessary?
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: You did comment out all uses, so I'd say remove it!

File: arduino/WandCore/CAPClient.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+#include "CAPClient.h"
+
+// touch threshold
+const int touchThold = 50; //above means not touched
+
+// Touch surface timing variables
+int debounce = 20;                // ms debounce period to prevent flickering when pressing or releasing the button
+int DCgap = 250;                  // max ms between clicks for a double click event
+int holdTime = 1000;              // ms hold period: how long to wait for tap+hold event
```
Comment (rytrose) [inline arduino/WandCore/CAPClient.cpp]: My gut feeling says this can be a little shorter, but experimentation can resolve that!

File: arduino/WandCore/WandCore.ino [hunk 12]
```diff
@@ -78,14 +79,13 @@ float sensorB;
...
+// Touch Surface
 ////////////////////////////////////////////////////////////////////////////////////
+#include "CAPClient.h"
 
 // Pin used for capacitive touch button
-// NOTE: Touch pin T8 refers to physical pin 32, NOT 33 as it references in most
-//  documentation. This is due to some pin assignment issue, see:
-//  https://randomnerdtutorials.com/esp32-touch-pins-arduino-ide/
-int touchPin = T8;
+int touchPin = T9;
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: 🎉

File: arduino/WandCore/WandCore.ino [hunk 13]
```diff
@@ -96,6 +96,15 @@ int touchPin = T8;
 RemoteDebug Debug;
 ////////////////////////////////////////////////////////////////////////////////////
 
+////////////////////////////////////////////////////////////////////////////////////
+// Set up state
+////////////////////////////////////////////////////////////////////////////////////
+int wandMode = 1; // 1 = power/brightness, 2 = color
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: ...can we index by 0?

File: arduino/WandCore/WandCore.ino [hunk 14]
```diff
@@ -175,50 +190,135 @@ void setup()
...
+//  // Test color sensor
+//  setColorSensorLED(true);
+//  delay(1000);
+//  setColorSensorLED(false);
+//  for (int i = 0; i < 5; i++)
+//  {
+//    getRGB(&sensorR, &sensorG, &sensorB);
+//    delay(200);
+//  }
+//}
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: Why did you comment this out? If you don't want to use it, just comment out or remove its invocation.

File: arduino/WandCore/WandCore.ino [hunk 15]
```diff
@@ -143,11 +152,17 @@ void setup()
...
-      // KNOWN ISSUE - OTAing while telnet is open will cause the upload to fail. Need to implement throttling.
-      debugI("Progress: %d, Total: %d", progress, total);
+      int bigprog = 100 * progress;
+      int percent = bigprog / total;
+      int permod = percent % 10;
+      if (percent % 10 == 0 && percent > otaProgress) 
+      {
+        otaProgress = percent;
+        debugI("OTA upload is %d percent complete", percent);
+      }
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: I'm not super familiar with the logic here, can you add some comments? Also, I feel like `bigprog` is not a great variable name.

Comment (rytrose) [reply arduino/WandCore/WandCore.ino]: Does this get called for every int `progress`? (`1, 2, 3...`) I imagine this is in bytes.
If it doesn't then the `percent % 10 == 0` wouldn't always work, so I assume it does.

File: arduino/WandCore/WandCore.ino [hunk 16]
```diff
@@ -175,50 +190,135 @@ void setup()
...
+        setRGBLED(rainbowVar, 0, 255);
+        rainbowVar++;
+        if (rainbowVar == 255) rainbowSection++;
+        break;
+      case 6: // red stays up, green stays down, blue goes down
+        setRGBLED(255, 0, rainbowVar);
+        rainbowVar--;
+        if (rainbowVar == 0) rainbowSection = 1;
+        break;
+    }
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: This does work, but it's probably way easier to convert RBG to HSL and back and just cycle Hue through the hue spectrum (which with a constant Saturation and Lightness is a smooth color loop).

Here's a conversion implementation we could use: https://github.com/ratkins/RGBConverter

Comment (rytrose) [reply arduino/WandCore/WandCore.ino]: I also think this should be abstracted out into the `LEDClient`.
You could also set the speed, etc. This color loop implementation isn't necessarily going to have consistent speed, depending if iterations of `loop()` take different times.

File: arduino/WandCore/CAPClient.cpp (hunk 11 above)
Comment (zwrose) [reply arduino/WandCore/CAPClient.cpp]: noted, no changes now

File: arduino/WandCore/WandCore.ino (hunk 16 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino]: abstracting out now, but not updating the algorithm at this time, will mark as TODO

File: arduino/WandCore/CAPClient.cpp (hunk 9 above)
Comment (zwrose) [reply arduino/WandCore/CAPClient.cpp]: added lots o comments, but also simplified a lot as there was apparently a lot of cruft from the code snippet I pulled in.

File: arduino/WandCore/WandCore.ino (hunk 15 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino]: if you mod two ints and they don't divide evenly, it truncates the decimals off. added some comments and changed the variable name.

Comment (Ryan-Rose-Bose) [reply arduino/WandCore/WandCore.ino]: It's not about the modulo operation, it's about the value `percent`. This conditional logic is only called if `percent = 10`, `percent = 20`, `percent = 30`...
If this function is called after a random number of bytes are uploaded (e.g. `percent = 8`, `percent = 11`, `percent = 12`...), you may miss the case where `percent % 10 == 0`.

Comment (zwrose) [reply arduino/WandCore/WandCore.ino]: I see what you are saying in theory. In practice, this top level call (ArduinoOTA.onProgress) is really chatty, and what ends up happening is that there are like a dozen cases where `percent = 10` (or `20`, `30`, etc), hence the other logic of only actually logging it once. 

Worst case, it skips over one of the reports, which would be relatively fine imo.

Comment (Ryan-Rose-Bose) [reply arduino/WandCore/WandCore.ino]: Yeah, that's what I figured. Cool cool

File: arduino/WandCore/LEDClient.cpp [hunk 17]
```diff
@@ -10,6 +10,10 @@
 #define CHAN_1 1
 #define CHAN_2 2
 
+// RGB Colorloop state vars
+int RGBColorloopSection = 0; // this holds the state of where in the 6 rainbow sections (0-5) the colorloop is
```
Comment (rytrose) [inline arduino/WandCore/LEDClient.cpp]: Nit-picky thing, usually in camelCase conventional languages, even initialisms get lowercased if they're the first word, so this would be `rgbColorloopSection`.

File: arduino/WandCore/LEDClient.h [hunk 18]
```diff
@@ -1,18 +1,21 @@
...
+
+// Initializes RGB pins.
+void setupRGBLED(int redPin, int greenPin, int bluePin);
+
+// Sets the LED to an RGB color.
+void setRGBLED(int r, int g, int b);
+
+// Sets the LED to the next RGB color in the colorloop.
+void incrementRGBColorloop();
+
```
Comment (rytrose) [inline arduino/WandCore/LEDClient.h:20]: Do you think this begs for a `resetRGBColorloop()` to reset the colorloop state variables?

File: arduino/WandCore/TouchSurfaceClient.cpp [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+*/
+
+#include "TouchSurfaceClient.h"
+
+// Pin the touch surface is attached to
+int touchPin;
+
+// Touch surface constants
+const int touchThold = 50;                  // readings above mean the sensor is not being touched
+const int debounce = 20;                    // ms debounce period to prevent flickering when touching or stopping to touch the surface
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.cpp]: > or stopping to touch the surface

Maybe word this as "releasing from the surface"?

File: arduino/WandCore/TouchSurfaceClient.cpp [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+  Manages a Touch Surface Input
+*/
+
+#include "TouchSurfaceClient.h"
+
+// Pin the touch surface is attached to
+int touchPin;
+
+// Touch surface constants
+const int touchThold = 50;                  // readings above mean the sensor is not being touched
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.cpp]: I read this as `touch` `T` `hold` when I first saw it, I think I would name it `touchThresh` or `touchThreshold`.

File: arduino/WandCore/TouchSurfaceClient.cpp [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+const int debounce = 20;                    // ms debounce period to prevent flickering when touching or stopping to touch the surface
+const int doubleTapGap = 250;               // max ms between taps for a double tap event
+const int holdTime = 1000;                  // ms hold period: how long to wait for hold event; TODO experiment if this can shorten
+
+// Touch surface state variables
+boolean isTouched = false;                  // whether or not the touch surface is being touched during this loop
+boolean wasTouchedLast = false;             // whether or not the touch surface was touched last loop
+boolean doubleTapWaiting = false;           // true if a double tap may be pending
+boolean doubleTapOnUp = false;              // true if double tap conditions have been met but the second tap is still being touched
+unsigned long downTime = -1;                // most recent time when the surface was first touched
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.cpp]: > // most recent time when the surface was first touched

I think _"first"_ in this comment is not correct. It looks like you still update `downTime` on the second touch of a double tap.

File: arduino/WandCore/TouchSurfaceClient.cpp [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+  && wasTouchedLast == false                // last time through it did NOT sense a touch
+  && (millis() - upTime) > debounce) {      // debounce
+    downTime = millis();                    // set the new time of the most recent touch    
+    
+    if ((millis() - upTime) < doubleTapGap  // the time since the previous release is less than the double tap gap
+    && doubleTapOnUp == false               // we aren't already planning to send a double tap event once no longer being touched
+    && doubleTapWaiting == true) {          // we were waiting to see if it was a double tap (and not just a single tap)
+      doubleTapOnUp = true;                 // this is a double tap! send the event when no longer being touched
+      doubleTapWaiting = false;             // reset this since we're no longer waiting and seeing
+    } else  doubleTapOnUp = false;          // otherwise it's not a double tap
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.cpp]: There's an extra space between `else` and `doubleTapOnUp` here. \*twitches\*

File: arduino/WandCore/TouchSurfaceClient.cpp (hunk 19 above)
Comment (rytrose) [reply arduino/WandCore/TouchSurfaceClient.cpp]: If you want to be consistent with later uses, maybe "stopping of touching the surface".

File: arduino/WandCore/TouchSurfaceClient.cpp [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+    // first time through, set vars to ensure clean completion of hold
+    if (not holdEventOngoing) {             // previously this was not a hold event
+      doubleTapOnUp = false;                // might have been a double tap leading into the hold, but that should be canceled now
+      doubleTapWaiting = false;             // no longer waiting for a double tap as we're in a hold
+      holdEventOngoing = true;              // we're in a hold event now
+    }
+  }
+  wasTouchedLast = isTouched;               // store the state for the next pass
+  return event;                             
+}
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.cpp:115]: This function is so much better! :)

File: arduino/WandCore/TouchSurfaceClient.h [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+
+  Manages a Touch Surface Input
+*/
+
+#ifndef TOUCH_CLIENT
+#define TOUCH_CLIENT
+
+#include "Arduino.h"
+
+// definitions of touch events
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.h]: Let's be consistent with casing.

File: arduino/WandCore/TouchSurfaceClient.h [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+#define NO_INTERACTION 0
+#define SINGLE_TAP 1
+#define DOUBLE_TAP 2
+#define HOLDING 3
+
+//Establishes the pin that the touch surface is attached to
```
Comment (rytrose) [inline arduino/WandCore/TouchSurfaceClient.h]: And with spaces!

File: arduino/WandCore/WandCore.ino [hunk 26]
```diff
@@ -99,19 +100,19 @@ RemoteDebug Debug;
 ////////////////////////////////////////////////////////////////////////////////////
 // Set up state
 ////////////////////////////////////////////////////////////////////////////////////
-int wandMode = 1; // 1 = power/brightness, 2 = color
-int rainbowSection = 1; // this holds the state of where in the 6 rainbow sections the colorloop is
-int rainbowVar = 0; // this holds the state of the rgb colorloop variable
+// wand mode definitions
+#define POWER 0
+#define COLOR 1
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino:111]: Usually definitions are put at the top of a file, or in a header file, but I don't think there's necessarily anything wrong with doing it here. Just noting.

File: arduino/WandCore/WandCore.ino [hunk 27]
```diff
@@ -212,110 +213,75 @@ void setup()
...
+  delay(1000);
+  setRGBLED(0, 0, 0);
+  delay(1000);
+
+  // Test color sensor
+  setColorSensorLED(true);
+  delay(1000);
+  setColorSensorLED(false);
+  for (int i = 0; i < 5; i++)
+  {
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: I think I did this, but if you want, I'd put this `{` on the same line as the `for`.

File: arduino/WandCore/WandCore.ino [hunk 28]
```diff
@@ -99,19 +100,19 @@ RemoteDebug Debug;
 ////////////////////////////////////////////////////////////////////////////////////
 // Set up state
 ////////////////////////////////////////////////////////////////////////////////////
-int wandMode = 1; // 1 = power/brightness, 2 = color
-int rainbowSection = 1; // this holds the state of where in the 6 rainbow sections the colorloop is
-int rainbowVar = 0; // this holds the state of the rgb colorloop variable
+// wand mode definitions
+#define POWER 0
+#define COLOR 1
+int wandMode = POWER; // default on boot
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino:112]: We may want a `lastWandMode`, too, in order to handle transitions if we want.

File: arduino/WandCore/WandCore.ino [hunk 29]
```diff
@@ -330,23 +296,32 @@ void loop()
...
+      if (wandMode == POWER) myHue.setLightPower(lightID, myHue.OFF); // turn off the light
+      else if (wandMode == COLOR) { // toggle colorloop on bulb
+        if (isLightColorloopOn) {
+          myHue.setLightColorloop(lightID, myHue.OFF);
+          isLightColorloopOn = false;
+        }
+        else {
+          myHue.setLightColorloop(lightID, myHue.ON);
+          isLightColorloopOn = true;
+        }
```
Comment (rytrose) [inline arduino/WandCore/WandCore.ino]: Not that I think it's that big of a deal, but this may get out of sync with reality. We don't have any reading of Hue state in here yet, not sure if we want to add it.

**Review by rytrose — COMMENTED** [review]
Awesome stuff!

File: arduino/WandCore/LEDClient.h (hunk 18 above)
Comment (zwrose) [reply arduino/WandCore/LEDClient.h:20]: I considered this as well, but ultimately decided that there's no particular reason why the colorloop needs to start back at red each time, so in the name of KISS i opted not to include that.

File: arduino/WandCore/TouchSurfaceClient.cpp (hunk 21 above)
Comment (zwrose) [reply arduino/WandCore/TouchSurfaceClient.cpp]: yeah first isn't the right word here, I intended it to mean "started to be touched", not relating at all to single or double tap. Will tweak.

File: arduino/WandCore/WandCore.ino (hunk 26 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:111]: ok - i'm going to keep it here because I like the logical grouping.

File: arduino/WandCore/WandCore.ino (hunk 28 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:112]: nothing immediately comes to mind as to what we'd want to handle, so i'm just going to leave it out for now. simple enough to add later.

File: arduino/WandCore/WandCore.ino (hunk 29 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino]: yeah, we do need to add state I realized today as it's vital for incrementing/decrementing brightness. Will update this once I've implemented getting of state

File: arduino/WandCore/WandCore.ino [hunk 30]
```diff
@@ -163,7 +168,7 @@ void setup()
       }
     })
     .onError([](ota_error_t error) {
-      debugE("OTA Error[%d]: ", error);
+      debugE("OTA Error[%s]: ", error);
```
Comment (Ryan-Rose-Bose) [inline arduino/WandCore/WandCore.ino:171]: I haven't looked, but I'm guessing the type `ota_error_t` isn't a string, and your previous `%d` was probably more valid. Generally these error types are enumerations (like you check against below) and are often unsigned ints.

File: arduino/WandCore/WandCore.ino [hunk 31]
```diff
@@ -253,72 +257,114 @@ void loop()
...
-      break;
-    case HOLDING:
-      // TODO adjust brightness or color depending on mode
-      if ((millis() - lastLogged) >= 500) { // DEBUG ONLY, throttling log output
-        debugI("touch is being held");
-        lastLogged = millis();
-      }
+  if (currentTouchEvent == SINGLE_TAP) {
+    debugI("Single Tap Just Happened");
+    if (wandMode == POWER) {                      // power mode, single tap == set light brightness based on "brightness" of color sensed
```
Comment (Ryan-Rose-Bose) [inline arduino/WandCore/WandCore.ino:268]: > // power mode, single tap == set light brightness based on "brightness" of color sensed

I see you also put in the Miro "define what this means", and I'm not sure there is a good definition? I'm not sure we need to have a single tap function in this state.

File: arduino/WandCore/WandCore.ino [hunk 32]
```diff
@@ -253,72 +257,114 @@ void loop()
...
     readAndPredictGesture(imuData, gestureConfidence);
     
-    if (gestureConfidence[GESTURE_FLICK] > 0.9) { // it's a flick, do flick things!
+    // read target light state
+    // only do this here as it's not needed elsewhere and this minimizes polling
+    // step 1 of 3 - prepare the raw string from the Hue library
+    String rawLightState;
+    rawLightState = myHue.getLightInfo(lightID);
+    int removeToIdx = rawLightState.indexOf("{") - 1;
+    rawLightState.remove(0, removeToIdx);
```
Comment (Ryan-Rose-Bose) [inline arduino/WandCore/WandCore.ino:304]: Why is this needed?

File: arduino/WandCore/WandCore.ino [hunk 33]
```diff
@@ -253,72 +257,114 @@ void loop()
...
+    StaticJsonDocument<112> filter;
+    JsonObject filter_state = filter.createNestedObject("state");
+    filter_state["hue"] = true;
+    filter_state["on"] = true;
+    filter_state["effect"] = true;
+    filter_state["bri"] = true;
+    filter_state["sat"] = true;
+    filter_state["ct"] = true;
+    
+    StaticJsonDocument<192> doc;
```
Comment (Ryan-Rose-Bose) [inline arduino/WandCore/WandCore.ino:316]: Can you comment on the magic numbers `112` and `192` here?

File: arduino/WandCore/WandCore.ino [hunk 34]
```diff
@@ -253,72 +257,114 @@ void loop()
...
+    DeserializationError error = deserializeJson(doc, rawLightState, DeserializationOption::Filter(filter));
+    
+    if (error) {
+      debugE("deserializeJson failed: %s", error.c_str());
+      return;
+    }
+
+    // step 3 of 3 - assign temporary state variables
+    JsonObject lightCurrentState = doc["state"];
+    long lightCurrentHue = lightCurrentState["hue"]; // 0(red)-65535(red); green = 21845 and blue = 43690
```
Comment (Ryan-Rose-Bose) [inline arduino/WandCore/WandCore.ino:327]: You're going to need an HSL to RGB converter anyway if we want to shadow the color on the wand RGB :)

File: arduino/WandCore/WandCore.ino [hunk 35]
```diff
@@ -253,72 +257,114 @@ void loop()
...
   float gSum = fabs(imuData[GX]) + fabs(imuData[GY]) + fabs(imuData[GZ]);
 
-  // Acceleration threshold reached, predict gesture
+  // Acceleration threshold reached, handle gesture
   if (aSum >= accelerationThresholdG || gSum >= gyroThreshold) {
+    // do the gesture predicting
     readAndPredictGesture(imuData, gestureConfidence);
     
-    if (gestureConfidence[GESTURE_FLICK] > 0.9) { // it's a flick, do flick things!
+    // read target light state
```
Comment (Ryan-Rose-Bose) [inline arduino/WandCore/WandCore.ino:298]: This whole interaction with Hue should be abstracted out into a new function.

File: arduino/WandCore/WandCore.ino (hunk 31 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:268]: We could do nothing, though I think we could also use Saturation or Value as a proxy for brightness relatively easily? i just hadn't picked something yet. would need to do an RGB to HSV conversion.

File: arduino/WandCore/WandCore.ino (hunk 32 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:304]: because the Hue client provides a string of literally the entire HTTP response, including all the headers etc. This cuts the string down to just the JSON string.

I can add comments to capture this in the code.

File: arduino/WandCore/WandCore.ino (hunk 31 above)
Comment (rytrose) [reply arduino/WandCore/WandCore.ino:268]: We don't get anything other than hue information from the color sensor -- saturation and brightness/value are independent of hue. In any case we're dealing with, S and V/L in an RGB<->HSV(L) conversion will be fixed.

File: arduino/WandCore/WandCore.ino (hunk 33 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:316]: will add a comment in the code, but they come from using the ArduinoJson assistant https://arduinojson.org/v6/assistant/ that recommends the size of these documents based on examples you feed it. Generated nifty custom sample code too!

File: arduino/WandCore/WandCore.ino (hunk 34 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:327]: yeahhh we're going to need to do some converting lol

File: arduino/WandCore/WandCore.ino (hunk 33 above)
Comment (rytrose) [reply arduino/WandCore/WandCore.ino:316]: Ooooh dope, I had never seen that!

File: arduino/WandCore/WandCore.ino (hunk 35 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:298]: yeah i figured you might say that, though I'm not really sure what gets passed in and out? Should we set the state variables globally and just make it a void? Or should we create a buffer in loop that holds the vars and gets passed in, similar to what we do with IMU data?

Also, where does this abstraction live -- above setup()?

File: arduino/WandCore/WandCore.ino (hunk 31 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:268]: hmm. doesn't the "clear" value from the sensor influence RGB in some way? or otherwise give us a dimension beside hue?

File: arduino/WandCore/WandCore.ino (hunk 35 above)
Comment (rytrose) [reply arduino/WandCore/WandCore.ino:298]: Maybe its own file, I think it's really a separate set of functionality. Input is the light ID, output is the doc of that light's state, I would think.

File: arduino/WandCore/WandCore.ino (hunk 31 above)
Comment (rytrose) [reply arduino/WandCore/WandCore.ino:268]: Oh, I see, that was really buried in the docs/API. I suppose it does give a value of intensity, but I'm still not sure it'd be that worthwhile. Practical use of this seems iffy, what's a good usecase?

File: arduino/WandCore/WandCore.ino (hunk 30 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:171]: yep, did some more homework and you're right, this was an overcorrect. switching back!

File: arduino/WandCore/WandCore.ino (hunk 31 above)
Comment (zwrose) [reply arduino/WandCore/WandCore.ino:268]: lol it's a toy wand, not sure "practical" is a primary lens here; that said, it's not the most compelling, so we can just scratch it. the use case would be touching a darker color (e.g. black, dark blue) and setting a dim brightness, or a lighter color (white, light pink, etc) and setting a higher brightness.

Comment (rytrose) [reply arduino/WandCore/WandCore.ino:268]: I meant practical in the sense that we probably don't want the wand to have behaviors that are unclear, especially because it is a toy and a 3yo is the target audience! 

I haven't seen the clear sensor in action, but from reading I don't think that's how it works. It's a separate sensor from RGB, so a light source (regardless of black / white color) would read a higher value than not a light source (even if it was white, light pink, etc.). And we don't have good brightness control of the wand RGB LED so we couldn't display feedback very well.
